### ISSUE TITLE
[Research] GRU-Based Sleep Staging with Permutation Channel Selection for 4-Channel EEG

### ISSUE THREAD

**Issue by LakshmiSravyaVedantham** [LakshmiSravyaVedantham/DreamAnalysis#31]
## Paper
**Title:** GRU-powered sleep stage classification with permutation-based EEG channel selection
**Authors:** Ito et al. (University of Tsukuba / WPI-IIIS)
**Year:** 2024
**Venue:** Scientific Reports
**URL:** https://www.nature.com/articles/s41598-024-68978-4

## Key Contribution
Introduces a computationally inexpensive permutation-based method for EEG channel selection combined with GRU deep learning models for 5-class sleep staging. Key finding: **as few as 3 randomly selected channels by permutation can match or beat the 3 AASM-recommended channels**, and performance degrades drastically below 3 channels. This validates that Muse 2's 4 channels (TP9, AF7, AF8, TP10) are sufficient for accurate sleep staging.

## Relevance to NeuralDreamWorkshop
Current `sleep_staging.py` uses an EEGNet variant with feature-based fallback (92.98% accuracy). A GRU architecture optimized for few-channel EEG could improve accuracy and provide better temporal modeling across epochs — GRUs capture inter-epoch transitions (e.g., N2 → N3 → REM progressions) that per-epoch CNNs miss.

### Files to Create
- `ml/models/gru_sleep_staging.py` — GRU model class with bidirectional GRU layers
- `ml/training/train_gru_sleep.py` — Training script using ISRUC + Sleep-EDF datasets

### Files to Modify
- `ml/models/sleep_staging.py` — Add GRU as priority path above feature-based: ONNX → sklearn → **GRU** → feature-based
- `ml/api/routes.py` — Wire GRU model into `/sleep-stage` endpoint, add model type to response
- `ml/processing/eeg_processor.py` — Add `extract_sleep_sequence_features()` for multi-epoch temporal context (3-5 consecutive epochs as input)

### Architecture
```
Input: 4-ch EEG epoch (4 x 1024 samples @ 256Hz = 4 seconds)
  → Band power features per channel (5 bands x 4 channels = 20)
  → Temporal context: stack 3-5 consecutive epoch features
  → Bidirectional GRU (2 layers, hidden_size=64)
  → FC → Softmax → 5 classes (W, N1, N2, N3, REM)
```

### Permutation Channel Selection
- During training, systematically test all C(4,3)=4 combinations of 3 Muse channels
- Find optimal 3-channel subset for sleep staging (likely AF7+AF8+TP10 based on paper)
- Use this to gracefully degrade when one channel has poor contact

## Expected Impact
- **Accuracy:** 92-95% (up from 92.98% — marginal but with better N1/REM discrimination)
- **Key win:** Better N1 detection (the hardest stage) through temporal context modeling
- **Graceful degradation:** Known fallback when 1 channel has artifacts

## Complexity: MEDIUM
- GRU architecture is straightforward in PyTorch (~200 lines)
- Training requires ISRUC dataset (free, 10 subjects)
- Main challenge: building the multi-epoch sliding window pipeline
